examples/raytracing/example_raytracing.cpp

```diff
@@ -119,83 +194,141 @@ int main(int argc, char *argv[])
+
+  // For every box shoot rays from random (uniformly distributed) points inside
+  // the box in random (uniformly distributed) directions.
+  Kokkos::Profiling::pushRegion("make_rays");
+  Kokkos::View<ArborX::Experimental::Ray *, MemorySpace> rays(
+      Kokkos::view_alloc(Kokkos::WithoutInitializing, "rays"),
+      num_rays * num_boxes);
+  {
+    using RandPoolType = Kokkos::Random_XorShift64_Pool<>;
+    RandPoolType rand_pool(5374857);
+    using GeneratorType = RandPoolType::generator_type;
+
+    Kokkos::parallel_for(
+        "initialize_rays",
+        Kokkos::MDRangePolicy<Kokkos::Rank<2>, ExecutionSpace>(
+            exec_space, {0, 0}, {num_boxes, num_rays}),
+        KOKKOS_LAMBDA(const size_t i, const size_t j) {
+          GeneratorType random_generator = rand_pool.get_state();
+          using Kokkos::Experimental::cos;
+          using Kokkos::Experimental::sin;
+          using Kokkos::Experimental::acos;
+
+          float xi_1 =
+              Kokkos::rand<GeneratorType, float>::draw(random_generator);
+          float xi_2 =
+              Kokkos::rand<GeneratorType, float>::draw(random_generator);
+          float xi_3 =
+              Kokkos::rand<GeneratorType, float>::draw(random_generator);
+          float xi_4 =
+              Kokkos::rand<GeneratorType, float>::draw(random_generator);
+          float xi_5 =
+              Kokkos::rand<GeneratorType, float>::draw(random_generator);
+
+          float upsilon = 2 * M_PI * xi_1;
+          float theta = acos(1 - 2 * xi_2);
+
+          rays(j + i * num_rays) = {{xi_3 * dx + boxes(i).minCorner()[0],
+                                     xi_4 * dy + boxes(i).minCorner()[1],
+                                     xi_5 * dz + boxes(i).minCorner()[2]},
+                                    {cos(upsilon) * sin(theta),
+                                     sin(upsilon) * sin(theta), cos(theta)}};
+
+          rand_pool.free_state(random_generator);
+        });
+  }
+  Kokkos::Profiling::popRegion();
+  Kokkos::Profiling::popRegion();
+
+  // Construct BVH
+  ArborX::BVH<MemorySpace> bvh{exec_space, boxes};
+
+  // Trace Rays
+  Kokkos::View<float *, MemorySpace> ray_energy(
+      Kokkos::view_alloc("ray_energy", Kokkos::WithoutInitializing),
+      num_rays * num_boxes);
+  Kokkos::deep_copy(ray_energy, (4000. * dx * dy * dz) / num_rays);
+  Kokkos::View<float *, MemorySpace> my_energy("energy", num_boxes);
+
+  Kokkos::Profiling::pushRegion("first_approach");
+  bvh.query(exec_space, RaysFirst<MemorySpace>{rays},
+            DepositEnergy<MemorySpace>{boxes, ray_energy, my_energy});
+  Kokkos::Profiling::popRegion();
+
+  Kokkos::Profiling::pushRegion("second_approach");
+  Kokkos::View<IntersectedCell *> values("values", 0);
+  Kokkos::View<int *> offsets("offsets", 0);
+  bvh.query(exec_space, RaysSecond<MemorySpace>{rays},
+            AccumRaySphereOptDist<MemorySpace>{boxes}, values, offsets);
+  ArborX::Details::sortObjects(exec_space, values);
```

Comment:
For a better comparison, one should create a vector of `long long` or `double` with each entry encoding a combination of `key` and `rid`. Then use it to produce a permutation. This would run about 2.5x faster than the sort on these structs with 4 values.

Reply:
Using a struct with only the required variables for sorting, shaves off like 0.5s. For a total of 1.13s for the original approach (vs 0.42s for the new approach).

 | |old | new |
| -- | -- | -- |
| sorting | 1.08e+00s  | 4.45e-01s |
| deposit_energy | 8.62e-02s |  1.55e-01s |
| total | 1.66se+00s | 1.07e+00s |